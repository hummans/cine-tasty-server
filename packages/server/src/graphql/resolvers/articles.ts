import { QueryResolvers } from '../../lib/types';
import { Datasource } from '../datasources';
import { Article } from '../../lib/types';

const resolvers: QueryResolvers = {
  Query: {
    articles: (
      _: {},
      __: {},
      { dataSources }: Datasource,
    ): Promise<Article[]> => dataSources.news.getAllArticles(),
  },
};

export default resolvers;
