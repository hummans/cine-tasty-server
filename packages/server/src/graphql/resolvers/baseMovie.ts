import MediaGenresHandler from '../datasources/the-movie-db-api/handlers/media-genres';
import {
  BaseMovieGenre_IdsArgs as BaseMovieGenreIdsArgs,
  QueryResolvers,
  MediaType,
} from '../../lib/types';
import { MediaItem } from '../../types';

const mediaGenres = new MediaGenresHandler();

const resolvers: QueryResolvers = {
  BaseMovie: {
    genre_ids: (
      { genre_ids }: MediaItem,
      { language }: BaseMovieGenreIdsArgs,
    ): Promise<string[]> =>
      mediaGenres.getMediaGenres(genre_ids, MediaType.Movie.toLowerCase(), language),
  },
};

export default resolvers;
