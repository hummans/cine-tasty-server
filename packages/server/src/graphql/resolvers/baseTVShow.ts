import MediaGenresHandler from '../datasources/the-movie-db-api/handlers/media-genres';
import {
  BaseTvShowGenre_IdsArgs as BaseTvShowGenreIdsArgs,
  QueryResolvers,
  MediaType,
} from '../../lib/types';
import { MediaItem } from '../../types';

const mediaGenres = new MediaGenresHandler();

const resolvers: QueryResolvers = {
  BaseTVShow: {
    genre_ids: (
      { genre_ids }: MediaItem,
      { language }: BaseTvShowGenreIdsArgs,
    ): Promise<string[]> =>
      mediaGenres.getMediaGenres(genre_ids, MediaType.Tv.toLowerCase(), language),
  },
};

export default resolvers;
