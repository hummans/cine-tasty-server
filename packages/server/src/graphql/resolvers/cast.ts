import { QueryResolvers, MediaType } from '../../lib/types';

type CastType = {
  media_type: 'tv' | 'movie';
};

const resolveCastType = (cast: CastType): string | null => {
  if (cast.media_type === MediaType.Movie.toLowerCase()) {
    return 'CastMovie';
  }

  if (cast.media_type === MediaType.Tv.toLowerCase()) {
    return 'CastTV';
  }

  return null;
};

const resolveTypes: QueryResolvers = {
  Cast: {
    __resolveType(cast: CastType): string | null {
      return resolveCastType(cast);
    },
  },
};

export default resolveTypes;
