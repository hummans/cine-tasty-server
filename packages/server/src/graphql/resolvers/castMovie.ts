import { MediaItem } from '../../types';
import {
  QueryResolvers,
  MediaType,
  CastMovieGenre_IdsArgs as CastMovieGenreIdsArgs,
} from '../../lib/types';
import MediaGenresHandler from '../datasources/the-movie-db-api/handlers/media-genres';

const mediaGenres = new MediaGenresHandler();

const resolvers: QueryResolvers = {
  CastMovie: {
    genre_ids: (
      { genre_ids }: MediaItem,
      { language }: CastMovieGenreIdsArgs,
    ): Promise<string[]> => {
      return mediaGenres.getMediaGenres(
        genre_ids,
        MediaType.Movie.toLowerCase(),
        language,
      );
    },
  },
};

export default resolvers;
