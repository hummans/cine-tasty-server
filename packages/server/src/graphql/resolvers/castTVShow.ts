import { MediaItem } from '../../types';
import {
  QueryResolvers,
  MediaType,
  CastTvShowGenre_IdsArgs as CastTvGenreIdsArgs,
} from '../../lib/types';
import MediaGenresHandler from '../datasources/the-movie-db-api/handlers/media-genres';

const mediaGenres = new MediaGenresHandler();

const resolvers: QueryResolvers = {
  CastTVShow: {
    genre_ids: (
      { genre_ids }: MediaItem,
      { language }: CastTvGenreIdsArgs,
    ): Promise<string[]> => {
      return mediaGenres.getMediaGenres(genre_ids, MediaType.Tv.toLowerCase(), language);
    },
  },
};

export default resolvers;
