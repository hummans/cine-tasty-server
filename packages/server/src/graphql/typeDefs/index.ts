import ISO6391Language from './types/ISO6391Language';
import root from './root';
import productionCompany from './types/productionCompany';
import cast from './types/cast';
import mediaType from './types/mediaType';
import article from './types/article';
import person from './types/person';
import movie from './types/movie';
import search from './types/search';
import review from './types/review';
import tvshow from './types/tvshow';
import video from './types/mediaVideo';
import queries from './queries';

export default [
  root,
  movie,
  productionCompany,
  search,
  tvshow,
  review,
  video,
  article,
  cast,
  person,
  mediaType,
  person,
  ISO6391Language,
  queries,
];
