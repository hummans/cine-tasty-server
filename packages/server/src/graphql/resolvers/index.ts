import articles from './articles';
import baseMovie from './baseMovie';
import baseTVShow from './baseTVShow';
import cast from './cast';
import castMovie from './castMovie';
import castTV from './castTV';
import people from './people';
import person from './person';
import personKnownFor from './personKnownFor';
import search from './search';
import searchResultItem from './searchResultItem';

export default [
  articles,
  baseMovie,
  baseTVShow,
  cast,
  castTV,
  castMovie,
  people,
  person,
  personKnownFor,
  search,
  searchResultItem,
];
