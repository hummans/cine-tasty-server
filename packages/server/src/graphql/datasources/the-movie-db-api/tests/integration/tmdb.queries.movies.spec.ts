const mockRestDataSourceGet = jest.fn();

import { createTestClient } from 'apollo-server-testing';
import { ApolloServer, gql } from 'apollo-server';

import { movieGenres } from '../../../../../__tests__/mocks/mediaGenres.stub';
import { TrendingMoviesEndpoints } from '../../../../../types';
import {
  rawMovie,
  movie,
  rawMovieDetail,
  movieDetail,
  review,
} from '../../../../../__tests__/mocks/movies.stub';
import env from '../../../../../config/environment';
import resolvers from '../../../../resolvers';
import typeDefs from '../../../../typeDefs';
import TheMovieDBAPI from '../..';

const GENRE_MOVIE_ENDPOINT = '/genre/movie/list';

const GET_TRENDING_MOVIES = gql`
  fragment TrendingMovieItem on BaseMovie {
    genre_ids
    original_title
    video
    title
    adult
    release_date
    backdrop_path
    overview
    vote_average
    poster_path
    popularity
    original_language
    vote_count
    overview
    id
  }

  query TrendingMovies($page: Int!) {
    trending_movies {
      now_playing(args: { page: $page }) {
        total_results
        total_pages
        hasMore
        items {
          ...TrendingMovieItem
        }
      }
      popular(args: { page: $page }) {
        total_results
        total_pages
        hasMore
        items {
          ...TrendingMovieItem
        }
      }
      top_rated(args: { page: $page }) {
        total_results
        total_pages
        hasMore
        items {
          ...TrendingMovieItem
        }
      }
      upcoming(args: { page: $page }) {
        total_results
        total_pages
        hasMore
        items {
          ...TrendingMovieItem
        }
      }
    }
  }
`;

const GET_MOVIE_DETAIL = gql`
  query Movie($id: ID!, $language: ISO6391Language) {
    movie(id: $id, language: $language) {
      adult
      backdrop_path
      genres(language: $language)
      id
      original_language
      original_title
      overview
      poster_path
      popularity
      video
      title
      vote_average
      release_date
      production_companies {
        id
        logo_path
        name
        origin_country
      }
      vote_count
      runtime
      status
      tagline
      budget
      homepage
      revenue
      spoken_languages
      production_countries
      cast {
        name
        profile_path
        id
        character
      }
      crew {
        department
        id
        job
        name
        profile_path
      }
      videos {
        thumbnail {
          extra_small
          small
          medium
          large
          extra_large
        }
        key
        name
        site
        id
        type
      }
    }
  }
`;

const GET_MOVIE_REVIEWS = gql`
  query MovieReviews($id: ID!, $reviewsPage: Int!) {
    movie(id: $id) {
      reviews(id: $id, reviewsPage: $reviewsPage) {
        total_results
        total_pages
        hasMore
        items {
          author
          content
          id
          url
        }
      }
    }
  }
`;

const GET_MOVIE_SIMILARS = gql`
  query MovieSimilars($id: ID!, $similarsPage: Int!) {
    movie(id: $id) {
      similar(id: $id, similarsPage: $similarsPage) {
        total_results
        total_pages
        hasMore
        items {
          original_title
          video
          title
          adult
          release_date
          backdrop_path
          genre_ids
          overview
          vote_average
          poster_path
          popularity
          original_language
          vote_count
          id
        }
      }
    }
  }
`;

const makeTestServer = (): ApolloServer => {
  return new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({
      tmdb: new TheMovieDBAPI(),
    }),
  });
};

jest.mock('apollo-datasource-rest', () => {
  class MockRESTDataSource {
    baseUrl = '';
    get = mockRestDataSourceGet;
  }

  return {
    RESTDataSource: MockRESTDataSource,
    HTTPCache: class HTTPCache {},
  };
});

describe('Integration: DataSources-Movies', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Query - Similar Movies', () => {
    it('should query similar movies of a movie from TheMovieDB API and returns the result correctly', async () => {
      mockRestDataSourceGet
        .mockReturnValueOnce({})
        .mockReturnValueOnce({
          id: 1,
          page: 1,
          results: [rawMovie],
          total_pages: 1,
          total_results: 1,
        })
        .mockReturnValueOnce({ genres: movieGenres });

      const server = makeTestServer();

      const { query } = createTestClient(server);

      const { data } = await query({
        query: GET_MOVIE_SIMILARS,
        variables: { id: '1', similarsPage: 1 },
      });

      expect(mockRestDataSourceGet.mock.calls.length).toBe(3);

      expect(mockRestDataSourceGet).toHaveBeenCalledWith('movie/1', {
        append_to_response: 'videos,credits',
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
      });

      expect(mockRestDataSourceGet).toHaveBeenCalledWith('movie/1/similar', {
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
        page: 1,
      });

      expect(data!.movie.similar).toEqual({
        hasMore: false,
        total_pages: 1,
        total_results: 1,
        items: [movie],
      });
    });

    it('should query the similar movies of a movie from TheMovieDB API and returns the result correctly and returns hasMore field as true when has more items to be pagianted', async () => {
      mockRestDataSourceGet
        .mockReturnValueOnce({ similars: {} })
        .mockReturnValueOnce({
          id: 1,
          page: 1,
          results: [rawMovie],
          total_pages: 2,
          total_results: 2,
        })
        .mockReturnValueOnce({ genres: movieGenres });

      const server = makeTestServer();

      const { query } = createTestClient(server);

      const { data } = await query({
        query: GET_MOVIE_SIMILARS,
        variables: { id: '1', similarsPage: 1 },
      });

      expect(mockRestDataSourceGet.mock.calls.length).toBe(3);

      expect(mockRestDataSourceGet).toHaveBeenCalledWith('movie/1', {
        append_to_response: 'videos,credits',
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
      });

      expect(mockRestDataSourceGet).toHaveBeenCalledWith('movie/1/similar', {
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
        page: 1,
      });

      expect(data!.movie.similar).toEqual({
        hasMore: true,
        total_pages: 2,
        total_results: 2,
        items: [movie],
      });
    });
  });

  describe('Query - Movie Reviews', () => {
    it('should query the reviews of a movie from TheMovieDB API and returns the result correctly', async () => {
      mockRestDataSourceGet.mockReturnValueOnce({ reviews: {} }).mockReturnValueOnce({
        id: 1,
        page: 1,
        results: [review],
        total_pages: 1,
        total_results: 1,
      });

      const server = makeTestServer();

      const { query } = createTestClient(server);

      const { data } = await query({
        query: GET_MOVIE_REVIEWS,
        variables: { id: '1', reviewsPage: 1 },
      });

      expect(mockRestDataSourceGet.mock.calls.length).toBe(2);

      expect(mockRestDataSourceGet).toHaveBeenCalledWith('movie/1', {
        append_to_response: 'videos,credits',
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
      });

      expect(mockRestDataSourceGet).toHaveBeenCalledWith('movie/1/reviews', {
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
        page: 1,
      });

      expect(data!.movie.reviews).toMatchSnapshot();
    });

    it('fetches the reviews of a movie from TheMovieDB API and returns the result correctly and returns hasMore field as true when has more items to be pagianted', async () => {
      mockRestDataSourceGet.mockReturnValueOnce({ reviews: {} }).mockReturnValueOnce({
        id: 1,
        page: 1,
        results: [review],
        total_pages: 2,
        total_results: 2,
      });

      const server = makeTestServer();

      const { query } = createTestClient(server);

      const { data } = await query({
        query: GET_MOVIE_REVIEWS,
        variables: { id: '1', reviewsPage: 1 },
      });

      expect(mockRestDataSourceGet.mock.calls.length).toBe(2);

      expect(mockRestDataSourceGet).toHaveBeenCalledWith('movie/1', {
        append_to_response: 'videos,credits',
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
      });

      expect(mockRestDataSourceGet).toHaveBeenCalledWith('movie/1/reviews', {
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
        page: 1,
      });

      expect(data!.movie.reviews).toMatchSnapshot();
    });
  });

  describe('Query - Movie Detail', () => {
    it('should query the details of a movie from TheMovieDB API and returns the result correctly', async () => {
      mockRestDataSourceGet
        .mockReturnValueOnce(rawMovieDetail)
        .mockReturnValueOnce({ genres: movieGenres });

      const server = makeTestServer();

      const { query } = createTestClient(server);

      const { data } = await query({
        query: GET_MOVIE_DETAIL,
        variables: { id: 1, language: 'PTBR' },
      });

      expect(mockRestDataSourceGet.mock.calls.length).toBe(2);

      expect(mockRestDataSourceGet).toHaveBeenCalledWith('movie/1', {
        append_to_response: 'videos,credits',
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'pt-br',
      });

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(GENRE_MOVIE_ENDPOINT, {
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'pt-br',
      });

      expect(data!.movie).toEqual(movieDetail);
    });
  });

  describe('Query - Trending Movies', () => {
    it('should query the now playing/popular/top_rated/upcoming movies from TheMovieDB API and returns the result correctly', async () => {
      mockRestDataSourceGet
        .mockReturnValueOnce({
          total_pages: 1,
          total_results: 1,
          results: [rawMovie],
        })
        .mockReturnValueOnce({
          total_pages: 1,
          total_results: 1,
          results: [rawMovie],
        })
        .mockReturnValueOnce({
          total_pages: 1,
          total_results: 1,
          results: [rawMovie],
        })
        .mockReturnValueOnce({
          total_pages: 1,
          total_results: 1,
          results: [rawMovie],
        })
        .mockReturnValueOnce({ genres: movieGenres })
        .mockReturnValueOnce({ genres: movieGenres })
        .mockReturnValueOnce({ genres: movieGenres })
        .mockReturnValueOnce({ genres: movieGenres });

      const server = makeTestServer();

      const { query } = createTestClient(server);

      const { data } = await query({
        query: GET_TRENDING_MOVIES,
        variables: { page: 1 },
      });

      expect(mockRestDataSourceGet.mock.calls.length).toBe(8);

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(
        TrendingMoviesEndpoints.NowPlaying,
        {
          api_key: env.THE_MOVIE_DB_API_KEY,
          language: 'en-us',
          page: 1,
        },
      );

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(
        TrendingMoviesEndpoints.Popular,
        {
          api_key: env.THE_MOVIE_DB_API_KEY,
          language: 'en-us',
          page: 1,
        },
      );

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(
        TrendingMoviesEndpoints.TopRated,
        {
          api_key: env.THE_MOVIE_DB_API_KEY,
          language: 'en-us',
          page: 1,
        },
      );

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(
        TrendingMoviesEndpoints.Upcoming,
        {
          api_key: env.THE_MOVIE_DB_API_KEY,
          language: 'en-us',
          page: 1,
        },
      );

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(GENRE_MOVIE_ENDPOINT, {
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
      });

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(GENRE_MOVIE_ENDPOINT, {
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
      });

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(GENRE_MOVIE_ENDPOINT, {
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
      });

      expect(mockRestDataSourceGet).toHaveBeenCalledWith(GENRE_MOVIE_ENDPOINT, {
        api_key: env.THE_MOVIE_DB_API_KEY,
        language: 'en-us',
      });

      expect(data!.trending_movies).toMatchSnapshot();
    });
  });
});
