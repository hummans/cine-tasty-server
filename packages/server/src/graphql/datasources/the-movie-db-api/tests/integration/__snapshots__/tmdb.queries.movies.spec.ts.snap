// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: DataSources-Movies Query - Trending Movies should query the now playing/popular/top_rated/upcoming movies from TheMovieDB API and returns the result correctly 1`] = `
Object {
  "now_playing": Object {
    "hasMore": false,
    "items": Array [
      Object {
        "adult": false,
        "backdrop_path": "/backdrop_path.jpg",
        "genre_ids": Array [
          "Movie Genre",
        ],
        "id": 530915,
        "original_language": "original_language",
        "original_title": "original_title",
        "overview": "overview",
        "popularity": 222.017,
        "poster_path": "/poster_path.jpg",
        "release_date": "release_date",
        "title": "title",
        "video": false,
        "vote_average": 8.1,
        "vote_count": 1530,
      },
    ],
    "total_pages": 1,
    "total_results": 1,
  },
  "popular": Object {
    "hasMore": false,
    "items": Array [
      Object {
        "adult": false,
        "backdrop_path": "/backdrop_path.jpg",
        "genre_ids": Array [
          "Movie Genre",
        ],
        "id": 530915,
        "original_language": "original_language",
        "original_title": "original_title",
        "overview": "overview",
        "popularity": 222.017,
        "poster_path": "/poster_path.jpg",
        "release_date": "release_date",
        "title": "title",
        "video": false,
        "vote_average": 8.1,
        "vote_count": 1530,
      },
    ],
    "total_pages": 1,
    "total_results": 1,
  },
  "top_rated": Object {
    "hasMore": false,
    "items": Array [
      Object {
        "adult": false,
        "backdrop_path": "/backdrop_path.jpg",
        "genre_ids": Array [
          "Movie Genre",
        ],
        "id": 530915,
        "original_language": "original_language",
        "original_title": "original_title",
        "overview": "overview",
        "popularity": 222.017,
        "poster_path": "/poster_path.jpg",
        "release_date": "release_date",
        "title": "title",
        "video": false,
        "vote_average": 8.1,
        "vote_count": 1530,
      },
    ],
    "total_pages": 1,
    "total_results": 1,
  },
  "upcoming": Object {
    "hasMore": false,
    "items": Array [
      Object {
        "adult": false,
        "backdrop_path": "/backdrop_path.jpg",
        "genre_ids": Array [
          "Movie Genre",
        ],
        "id": 530915,
        "original_language": "original_language",
        "original_title": "original_title",
        "overview": "overview",
        "popularity": 222.017,
        "poster_path": "/poster_path.jpg",
        "release_date": "release_date",
        "title": "title",
        "video": false,
        "vote_average": 8.1,
        "vote_count": 1530,
      },
    ],
    "total_pages": 1,
    "total_results": 1,
  },
}
`;
